File: suite/Constraints/Constraints.ml
exception EmptyList

type assign = Assign of int * int 

type csp = CSP of int * int * ((assign * assign) -> bool)

type 'a tree = Node of 'a * 'a tree list

type conflict_set = 
  | Known of int list 
  | Unknown

let value (Assign(_,value)) = value
let level (Assign(varr,_)) = varr

let label (Node(a,_)) = a

let rec map_list f ls = 
  match ls with
    | [] -> []
    | x::xs -> (f x)::(map_list f xs)

let rec all_list f ls = 
  match ls with 
    | [] -> true
    | x::xs -> (f x) && (all_list f xs)

let rec filter_list f ls = 
  match ls with 
    | [] -> []
    | x::xs -> if f x then x::(filter_list f xs) else filter_list f xs

let is_empty ls = 
  match ls with
    | [] -> true
    | _ -> false

let rec enum_from_to from to_ = 
  if from<=to_ then 
    from::(enum_from_to (from+1) to_)
  else []

let rec zip_with f x y = 
  match (x,y) with 
    | ([],_) -> []
    | (_,[]) -> []
    | (c::cs,p::ps) -> f(c,p) :: zip_with f cs ps

let rec len ls = 
  match ls with 
    | [] -> 0
    | _::xs -> 1 + (len xs)

let head ls = 
  match ls with 
    | [] -> raise EmptyList
    | x::_ -> x

let tail ls = 
  match ls with 
    | [] -> raise EmptyList
    | _::xs -> xs

let rec at_index ls ind = 
  match ls with 
    | [] -> raise EmptyList
    | x::xs -> if ind == 0 then x else at_index xs (ind-1)

let rec rev_loop ls acc = 
  match ls with 
    | [] -> acc
    | p::ps -> rev_loop ps (p::acc)

let reverse ls = rev_loop ls []

let rec concat_loop ls acc = 
  match ls with 
    | [] -> rev_loop acc []
    | l::ls -> concat_loop ls (rev_loop l acc)

let concat lss = concat_loop lss []

let rec append l1 l2 = 
  match l1 with 
    | [] -> l2
    | a::as_ -> a::(append as_ l2)

let rec foldl xs acc f = 
  match xs with 
    | [] -> acc
    | h::t -> foldl t (f acc h) f

let rec in_list i ls = 
  match ls with 
    | [] -> false
    | j::js -> if i == j then true else in_list i js

let not_elem a ls = not (in_list a ls)

let rec map_tree f (Node(l,ls)) = 
  Node(f l,map_list (fun x -> map_tree f x) ls)

let rec fold_tree f (Node(l,c)) = 
  f (l,map_list (fun x -> fold_tree f x) c)

let rec filter_tree p n =
  let f = fun (a,cs) ->
    Node(a,filter_list (fun x -> p (label x)) cs)
  in 
  fold_tree f n 

let rec leaves n = 
  match n with
    | Node(leaf,[]) -> leaf::[]
    | Node(leaf,cs) -> concat (map_list (fun x -> leaves x) cs)

let prune f n = 
  filter_tree (fun x -> not (f x)) n

let max_level ls = 
  match ls with 
    | [] -> 0
    | Assign(v,_)::_ -> v

let complete (CSP (v,_,_)) s = max_level s = v

let rec delete_by f x ys = 
  match ys with 
    | [] -> []
    | y::ys -> 
        if (f (x,y)) then ys
        else y::delete_by f x ys

let rec nub_by f l = 
  match l with 
    | [] -> []
    | h::t -> h::nub_by f (filter_list (fun y -> not (f(h,y))) t)

let union_by f l1 l2 = 
  append l1 
  (foldl
    (nub_by f l2)
    l1
    (fun acc -> fun y -> delete_by f y acc) 
  )

let union l1 l2 = union_by (fun (x,y) -> x=y) l1 l2

let rec combine ls acc = 
  match ls with 
    | [] -> acc
    | ((s,Known cs)::css) -> 
        if not_elem (max_level s) cs then
          cs 
        else 
          combine css (union cs acc)
    | ((s,Unknown)::_) -> acc

let rec init_tree f x = 
  Node(x, map_list (fun y -> init_tree f y) (f x))

let rec to_assign ls ss = 
  match ls with 
    | [] -> [] 
    | j::t1 -> (Assign((max_level ss) +1,j)::ss)::(to_assign t1 ss)


let mk_tree (CSP(vars,vals,_)) = 
  let next = fun ss -> 
    if max_level ss < vars then
      to_assign (enum_from_to 1 vals) ss
    else []
  in 
  init_tree next []


let rec collect ls = 
  match ls with 
    | [] -> []
    | Known cs :: css -> union cs (collect css)
    | Unknown :: _ -> []

let known_solution c = 
  match c with 
    | Known [] -> true
    | Known (_::_) -> false 
    | Unknown -> false

let known_conflict c = 
  match c with 
    | Known [] -> false 
    | Known (_::_) -> true
    | Unknown -> false

let rec filter_known ls = 
  match ls with 
    | [] -> []
    | vs::t1 -> if all_list known_conflict vs then 
      vs::(filter_known t1) 
    else 
      filter_known t1

let domain_wipeout csp t = 
  let f8 = fun ((as_,cs),tbl) ->
    let wiped_domains = filter_known tbl in
    let cs_ = 
      if is_empty wiped_domains then cs
      else Known (collect (head wiped_domains))
    in 
    (as_,cs_)
  in 
  map_tree f8 t 

let check_complete csp s = 
  if complete csp s then Known [] else Unknown

let earliest_inconsistency (CSP(_,_,rel)) aas = 
  match aas with 
    | [] -> None
    | a::as_ -> 
        (match (filter_list (fun x -> not (rel(a,x))) (reverse as_)) with
          | [] -> None
          | b::bs_ -> Some(level a, level b)
        )

let lookup_cache csp t = 
  let f5 = fun (csp, (ls,tbl)) -> 
    match ls with 
      | [] -> (([],Unknown),tbl)
      | a::as_ -> 
          let table_entry = at_index (head tbl) ((value a)-1) in 
          let cs = (
            match table_entry with 
              | Unknown -> check_complete csp (a::as_)
              | Known vals -> table_entry)
          in 
          ((a::as_,cs),tbl)
  in 
  map_tree (fun x -> f5 (csp,x)) t

let rec to_pairs ls varrr = 
  match ls with 
    | [] -> []
    | valll::t2 -> (varrr,valll) :: (to_pairs t2 varrr)


let rec n_pairs ls vals = 
  match ls with 
    | [] -> []
    | varrr::t1 -> 
        (to_pairs (enum_from_to 1 vals) varrr)::(n_pairs t1 vals)


let fill_table s (CSP(vars,vals,rel)) tbl = 
  match s with 
    | [] -> tbl
    | Assign (var_,val_)::as__ -> 
        let f4 = fun (cs,(varr,vall)) -> 
          (match cs with 
            | Known vs -> cs
            | Unknown -> 
                if not (rel (Assign(var_,val_),Assign(varr,vall))) 
                then Known (var_::varr::[])
                else cs)
        in 
        zip_with 
          (fun (x,y) -> zip_with f4 x y)
          tbl
          (n_pairs (enum_from_to (var_+1) vars) vals)

let rec cache_checks csp tbl (Node (s,cs)) = 
  Node ((s,tbl),
    map_list
      (fun x -> 
        cache_checks csp 
          (fill_table s csp (tail tbl)) 
        x) 
      cs)

let rec n_unknown ls n = 
  match ls with 
    | [] -> []
    | _::t1 -> 
        (to_unknown (enum_from_to 1 n)) 
        :: (n_unknown t1 n) 
and to_unknown ls = 
  match ls with
    | [] -> []
    | _::t2 -> Unknown::(to_unknown t2)
and empty_table (CSP(vars,vals,_)) = 
  []::(n_unknown (enum_from_to 1 vars) vals)

let search labeler csp = 
  map_list (fun (x,_) -> x)
    (filter_list (fun (_,x) -> known_solution x)
      (leaves 
        (prune (fun (_,x) -> known_conflict x)
          (labeler csp (mk_tree csp))
        )
      )
    )

let safe (Assign (i,m)) (Assign (j,n)) = 
  (not (m=n)) && (not ((abs (i-j)) = (abs (m-n))))

let queens n = CSP (n,n,fun (x,y) -> safe x y)

let bt csp t = 
  let f3 = fun s -> 
    match earliest_inconsistency csp s with 
      | Some (a,b) -> (s,Known (a::b::[]))
      | None -> (s,check_complete csp s)
  in 
  map_tree f3 t 

let bm csp t = 
  map_tree (fun (x,_) -> x)
    (lookup_cache csp 
      (cache_checks csp (empty_table csp) t))

let bj csp t = 
  let f6 = fun ((a,conf),chs) -> 
    match conf with 
      | Known cs -> Node((a,Known cs),chs)
      | Unknown -> 
          Node ((a,Known (combine (map_list (fun x -> label(x)) chs) [])), chs) 
  in
  fold_tree f6 t 

let bjbt csp t = bj csp (bt csp t)

let bj_ csp t = 
  let f6 = fun ((a,conf),chs) ->
    match conf with
      | Known cs -> Node ((a,Known cs),chs)
      | Unknown -> 
          let cs_ = Known (combine (map_list (fun x -> label x) chs) []) in 
          if known_conflict cs_ then 
            Node ((a,cs_),[])
          else Node((a,cs_),chs) in 
  fold_tree f6 t 

let bjbt_ csp t = bj_ csp (bt csp t)

let fc csp t = 
  domain_wipeout csp 
    (lookup_cache csp 
      (cache_checks csp (empty_table csp) t))

let try_ n algorithm = 
  len (search algorithm (queens n))

let test_constraints_nofib n = 
  map_list (fun x -> try_ n x) (bt::bm::bjbt::bjbt_::fc::[])

let rec main_loop iters n = 
  let res = test_constraints_nofib n in 
  if iters = 1 then
    print_endline (string_of_int (head res))
  else main_loop (iters-1) n 

let main = 
  let iters = int_of_string Sys.argv.(1) in 
  let n = int_of_string Sys.argv.(2) in 
  main_loop iters n

File: suite/Constraints/Constraints.rs
use std::rc::Rc;

#[derive(Debug, Clone)]
enum List<A> {
    Nil,
    Cons(A, Rc<List<A>>),
}

#[derive(Clone)]
struct Assign {
    varr: i64,
    value: i64,
}

struct CSP {
    vars: i64,
    vals: i64,
    rel: Box<dyn Fn(&Assign, &Assign) -> bool>,
}

#[derive(Clone)]
enum ConflictSet {
    Known(List<i64>),
    Unknown,
}

#[derive(Clone)]
struct Node<T> {
    lab: T,
    children: List<Rc<Node<T>>>,
}

impl<A> List<A> {
    fn map<B>(self, f: &impl Fn(A) -> B) -> List<B>
    where
        A: Clone,
    {
        match self {
            List::Nil => List::Nil,
            List::Cons(a, as_) => List::Cons(f(a), Rc::new(Rc::unwrap_or_clone(as_).map(f))),
        }
    }

    fn all(&self, f: &impl Fn(&A) -> bool) -> bool {
        match self {
            List::Nil => true,
            List::Cons(a, as_) => {
                if f(a) {
                    as_.all(f)
                } else {
                    false
                }
            }
        }
    }

    fn filter(self, f: &impl Fn(&A) -> bool) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => List::Nil,
            List::Cons(a, as_) => {
                if f(&a) {
                    List::Cons(a, Rc::new(Rc::unwrap_or_clone(as_).filter(f)))
                } else {
                    Rc::unwrap_or_clone(as_).filter(f)
                }
            }
        }
    }

    fn is_empty(&self) -> bool {
        matches!(self, List::Nil)
    }

    fn len(&self) -> usize {
        match self {
            List::Nil => 0,
            List::Cons(_, as_) => 1 + as_.len(),
        }
    }

    fn head(self) -> A {
        match self {
            List::Nil => panic!("Cannot take head of empty list"),
            List::Cons(a, _) => a,
        }
    }

    fn tail(self) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => panic!("Cannot take tail of empty list"),
            List::Cons(_, as_) => Rc::unwrap_or_clone(as_),
        }
    }

    fn at_index(self, n: usize) -> A
    where
        A: Clone,
    {
        match self {
            List::Nil => panic!("Cannot take {}th element of empty list", n),
            List::Cons(a, as_) => {
                if n == 0 {
                    a
                } else {
                    Rc::unwrap_or_clone(as_).at_index(n - 1)
                }
            }
        }
    }

    fn rev_loop(self, acc: List<A>) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => acc,
            List::Cons(p, ps) => Rc::unwrap_or_clone(ps).rev_loop(List::Cons(p, Rc::new(acc))),
        }
    }

    fn reverse(self) -> List<A>
    where
        A: Clone,
    {
        self.rev_loop(List::Nil)
    }

    fn append(self, other: List<A>) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => other,
            List::Cons(a, as_) => List::Cons(a, Rc::new(Rc::unwrap_or_clone(as_).append(other))),
        }
    }

    fn foldl(self, acc: List<A>, f: &impl Fn(List<A>, A) -> List<A>) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => acc,
            List::Cons(h, t) => Rc::unwrap_or_clone(t).foldl(f(acc, h), f),
        }
    }

    fn in_list(&self, a: &A) -> bool
    where
        A: PartialEq,
    {
        match self {
            List::Nil => false,
            List::Cons(a_, as_) => {
                if a_ == a {
                    true
                } else {
                    as_.in_list(a)
                }
            }
        }
    }

    fn not_elem(&self, a: &A) -> bool
    where
        A: PartialEq,
    {
        !self.in_list(a)
    }

    fn nub_by(self, f: impl Fn((A, A)) -> bool) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => List::Nil,
            List::Cons(h, t) => List::Cons(
                h.clone(),
                Rc::new(
                    Rc::unwrap_or_clone(t)
                        .filter(&|y| !f((h.clone(), y.clone())))
                        .nub_by(f),
                ),
            ),
        }
    }

    fn delete_by(self, x: A, f: impl Fn((A, A)) -> bool) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => List::Nil,
            List::Cons(y, ys) => {
                if f((x.clone(), y.clone())) {
                    Rc::unwrap_or_clone(ys)
                } else {
                    List::Cons(y, Rc::new(Rc::unwrap_or_clone(ys).delete_by(x, f)))
                }
            }
        }
    }

    fn union_by(self, other: List<A>, f: &impl Fn((A, A)) -> bool) -> List<A>
    where
        A: Clone + PartialEq,
    {
        self.clone().append(
            other
                .nub_by(f)
                .foldl(self, &|acc: List<A>, y| acc.delete_by(y, f)),
        )
    }

    fn union(self, other: List<A>) -> List<A>
    where
        A: Clone + PartialEq,
    {
        self.union_by(other, &|(x, y)| x == y)
    }

    fn zip<B>(self, other: List<B>) -> List<(A, B)>
    where
        A: Clone,
        B: Clone,
    {
        match (self, other) {
            (List::Nil, _) => List::Nil,
            (_, List::Nil) => List::Nil,
            (List::Cons(a, as_), List::Cons(b, bs_)) => List::Cons(
                (a, b),
                Rc::new(Rc::unwrap_or_clone(as_).zip(Rc::unwrap_or_clone(bs_))),
            ),
        }
    }

    fn zip_with<B, C>(self, other: List<B>, f: &impl Fn(A, B) -> C) -> List<C>
    where
        A: Clone,
        B: Clone,
    {
        self.zip(other).map(&|(x, y)| f(x, y))
    }
}

impl<A> List<List<A>>
where
    A: Clone,
{
    fn concat_loop(self, acc: List<A>) -> List<A> {
        match self {
            List::Nil => acc.rev_loop(List::Nil),
            List::Cons(l, ls) => Rc::unwrap_or_clone(ls).concat_loop(l.rev_loop(acc)),
        }
    }

    fn concat(self) -> List<A> {
        self.concat_loop(List::Nil)
    }
}

impl List<i64> {
    pub fn enum_from_to(from: i64, to: i64) -> List<i64> {
        if from <= to {
            List::Cons(from, Rc::new(List::enum_from_to(from + 1, to)))
        } else {
            List::Nil
        }
    }
}

impl<T, A> From<T> for List<A>
where
    T: Iterator<Item = A>,
{
    fn from(it: T) -> List<A> {
        let mut ls = List::Nil;
        for a in it {
            ls = List::Cons(a, Rc::new(ls));
        }
        ls
    }
}

impl<T> Node<T> {
    fn map<U>(self, f: &impl Fn(T) -> U) -> Node<U>
    where
        T: Clone,
    {
        Node {
            lab: f(self.lab),
            children: self
                .children
                .map(&|x| Rc::new(Rc::unwrap_or_clone(x).map(f))),
        }
    }

    fn fold<U>(self, f: &impl Fn(T, List<Node<U>>) -> Node<U>) -> Node<U>
    where
        T: Clone,
    {
        f(
            self.lab,
            self.children.map(&|x| Rc::unwrap_or_clone(x).fold(f)),
        )
    }

    fn filter(self, p: &impl Fn(&T) -> bool) -> Node<T>
    where
        T: Clone,
    {
        let f1 = |a, cs: List<Node<T>>| Node {
            lab: a,
            children: cs.filter(&|x| p(&x.lab)).map(&Rc::new),
        };
        self.fold(&f1)
    }

    fn leaves(self) -> List<T>
    where
        T: Clone,
    {
        match self.children {
            List::Nil => List::Cons(self.lab, Rc::new(List::Nil)),
            _ => self
                .children
                .map(&|x: Rc<Node<T>>| Rc::unwrap_or_clone(x).leaves())
                .concat(),
        }
    }

    fn prune(self, p: &impl Fn(&T) -> bool) -> Node<T>
    where
        T: Clone,
    {
        self.filter(&|x| !(p(x)))
    }
}

fn max_level(s: &List<Assign>) -> i64 {
    match s {
        List::Nil => 0,
        List::Cons(a, _) => a.varr,
    }
}

fn complete(csp: &CSP, s: &List<Assign>) -> bool {
    max_level(s) == csp.vars
}

fn combine(ls: List<(List<Assign>, ConflictSet)>, acc: List<i64>) -> List<i64> {
    match ls {
        List::Nil => acc,
        List::Cons((s, ConflictSet::Known(cs)), css) => {
            if cs.not_elem(&max_level(&s)) {
                cs
            } else {
                combine(Rc::unwrap_or_clone(css), cs.union(acc))
            }
        }
        List::Cons((_, ConflictSet::Unknown), _) => acc,
    }
}

fn init_tree(
    f: &impl Fn(List<Assign>) -> List<List<Assign>>,
    x: List<Assign>,
) -> Node<List<Assign>> {
    let children = f(x.clone()).map(&|y| Rc::new(init_tree(f, y)));
    Node { lab: x, children }
}

fn to_assign(ls: List<i64>, ass: List<Assign>) -> List<List<Assign>> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(j, t1) => List::Cons(
            List::Cons(
                Assign {
                    varr: max_level(&ass) + 1,
                    value: j,
                },
                Rc::new(ass.clone()),
            ),
            Rc::new(to_assign(Rc::unwrap_or_clone(t1), ass)),
        ),
    }
}

fn mk_tree(csp: &CSP) -> Node<List<Assign>> {
    let next = |ss: List<Assign>| {
        if max_level(&ss) < csp.vars {
            to_assign(List::enum_from_to(1, csp.vals), ss)
        } else {
            List::Nil
        }
    };
    init_tree(&next, List::Nil)
}

fn collect_conflict(ls: List<ConflictSet>) -> List<i64> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(ConflictSet::Known(cs), css) => {
            cs.union(collect_conflict(Rc::unwrap_or_clone(css)))
        }
        List::Cons(ConflictSet::Unknown, _) => List::Nil,
    }
}

fn filter_known(
    ls: List<List<ConflictSet>>,
    f: impl Fn(List<ConflictSet>) -> bool,
) -> List<List<ConflictSet>> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(vs, t1) => {
            if vs.all(&known_conflict) {
                List::Cons(vs, Rc::new(filter_known(Rc::unwrap_or_clone(t1), f)))
            } else {
                filter_known(Rc::unwrap_or_clone(t1), f)
            }
        }
    }
}

fn domain_wipeout(
    _: &CSP,
    t: Node<((List<Assign>, ConflictSet), List<List<ConflictSet>>)>,
) -> Node<(List<Assign>, ConflictSet)> {
    let f8 = |((as_, cs), tbl)| {
        let wiped_domains =
            filter_known(tbl, |vs: List<ConflictSet>| vs.all(&|x| known_conflict(x)));
        let cs_ = if wiped_domains.is_empty() {
            cs
        } else {
            let hd = wiped_domains.head();
            ConflictSet::Known(collect_conflict(hd))
        };
        (as_, cs_)
    };
    t.map(&f8)
}

fn check_complete(csp: &CSP, s: &List<Assign>) -> ConflictSet {
    if complete(csp, s) {
        ConflictSet::Known(List::Nil)
    } else {
        ConflictSet::Unknown
    }
}

fn earliest_inconsistency(csp: &CSP, aas: List<Assign>) -> Option<(i64, i64)> {
    match aas {
        List::Nil => None,
        List::Cons(a, aas_) => match Rc::unwrap_or_clone(aas_)
            .reverse()
            .filter(&|x| !((csp.rel)(&a, x)))
        {
            List::Nil => None,
            List::Cons(b, _) => Some((a.varr, b.varr)),
        },
    }
}

fn lookup_cache(
    csp: &CSP,
    t: Node<(List<Assign>, List<List<ConflictSet>>)>,
) -> Node<((List<Assign>, ConflictSet), List<List<ConflictSet>>)> {
    let f5 = |csp, (as_, tbl): (List<Assign>, List<List<ConflictSet>>)| match as_ {
        List::Nil => ((List::Nil, ConflictSet::Unknown), tbl),
        List::Cons(ref a, _) => {
            let table_entry = tbl.clone().head().at_index((a.value - 1) as usize);
            let cs = match table_entry {
                ConflictSet::Unknown => check_complete(csp, &as_),
                ConflictSet::Known(_) => table_entry.clone(),
            };
            ((as_, cs), tbl)
        }
    };
    t.map(&|x| f5(csp, x))
}

fn cache_checks(
    csp: &CSP,
    tbl: List<List<ConflictSet>>,
    n: Node<List<Assign>>,
) -> Node<(List<Assign>, List<List<ConflictSet>>)> {
    let s = n.lab.clone();
    Node {
        lab: (n.lab, tbl.clone()),
        children: n.children.map(&|x| {
            Rc::new(cache_checks(
                csp,
                fill_table(&s, csp, tbl.clone().tail()),
                Rc::unwrap_or_clone(x),
            ))
        }),
    }
}

fn known_solution(c: &ConflictSet) -> bool {
    match c {
        ConflictSet::Known(cs) => cs.is_empty(),
        ConflictSet::Unknown => false,
    }
}

fn known_conflict(c: &ConflictSet) -> bool {
    match c {
        ConflictSet::Known(cs) => !cs.is_empty(),
        ConflictSet::Unknown => false,
    }
}

fn to_pairs(ls: List<i64>, varrr: i64) -> List<(i64, i64)> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(valll, t2) => List::Cons(
            (varrr, valll),
            Rc::new(to_pairs(Rc::unwrap_or_clone(t2), varrr)),
        ),
    }
}

fn n_pairs(ls: List<i64>, n: i64) -> List<List<(i64, i64)>> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(varrr, t1) => List::Cons(
            to_pairs(List::enum_from_to(1, n), varrr),
            Rc::new(n_pairs(Rc::unwrap_or_clone(t1), n)),
        ),
    }
}

fn fill_table(
    s: &List<Assign>,
    csp: &CSP,
    tbl: List<List<ConflictSet>>,
) -> List<List<ConflictSet>> {
    match s {
        List::Nil => tbl,
        List::Cons(as_, _) => {
            let f4 = |cs, (varr, vall)| match cs {
                ConflictSet::Known(_) => cs,
                ConflictSet::Unknown => {
                    if !(csp.rel)(as_, &Assign { varr, value: vall }) {
                        ConflictSet::Known(List::Cons(
                            as_.varr,
                            Rc::new(List::Cons(varr, Rc::new(List::Nil))),
                        ))
                    } else {
                        cs
                    }
                }
            };
            tbl.zip_with(
                n_pairs(List::enum_from_to(as_.varr + 1, csp.vars), csp.vals),
                &|x, y| x.zip_with(y, &f4),
            )
        }
    }
}

fn to_unknown(ls: List<i64>) -> List<ConflictSet> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(_, t2) => List::Cons(
            ConflictSet::Unknown,
            Rc::new(to_unknown(Rc::unwrap_or_clone(t2))),
        ),
    }
}

fn n_unknown(ls: List<i64>, n: i64) -> List<List<ConflictSet>> {
    match ls {
        List::Nil => List::Nil,
        List::Cons(_, t1) => List::Cons(
            to_unknown(List::enum_from_to(1, n)),
            Rc::new(n_unknown(Rc::unwrap_or_clone(t1), n)),
        ),
    }
}

fn empty_table(csp: &CSP) -> List<List<ConflictSet>> {
    List::Cons(
        List::Nil,
        Rc::new(n_unknown(List::enum_from_to(1, csp.vars), csp.vals)),
    )
}

fn search(
    labeler: fn(&CSP, Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)>,
    csp: &CSP,
) -> List<List<Assign>> {
    let tree = mk_tree(csp);
    let labeled = labeler(csp, tree);
    let pruned = labeled.prune(&|(_, x)| known_conflict(&x));
    pruned
        .leaves()
        .filter(&|(_, x)| known_solution(x))
        .map(&|(x, _)| x)
}

fn safe(as1: &Assign, as2: &Assign) -> bool {
    (!(as1.value == as2.value)) && (!((as1.varr - as2.varr).abs() == (as1.value - as2.value).abs()))
}

fn queens(n: i64) -> CSP {
    CSP {
        vars: n,
        vals: n,
        rel: Box::new(safe),
    }
}

fn bt(csp: &CSP, t: Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)> {
    let f3 = |s: List<Assign>| {
        (
            s.clone(),
            match earliest_inconsistency(csp, s.clone()) {
                Some((a, b)) => {
                    ConflictSet::Known(List::Cons(a, Rc::new(List::Cons(b, Rc::new(List::Nil)))))
                }
                None => check_complete(csp, &s),
            },
        )
    };
    t.map(&f3)
}

fn bm(csp: &CSP, t: Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)> {
    lookup_cache(csp, cache_checks(csp, empty_table(csp), t)).map(&|(x, _)| x)
}

fn bj(_: &CSP, t: Node<(List<Assign>, ConflictSet)>) -> Node<(List<Assign>, ConflictSet)> {
    let f6 = |lp2, chs: List<Node<(List<Assign>, ConflictSet)>>| match lp2 {
        (a, ConflictSet::Known(cs)) => Node {
            lab: (a, ConflictSet::Known(cs)),
            children: chs.map(&Rc::new),
        },
        (a, ConflictSet::Unknown) => Node {
            lab: (
                a,
                ConflictSet::Known(combine(chs.clone().map(&|n| n.lab.clone()), List::Nil)),
            ),
            children: chs.map(&Rc::new),
        },
    };
    t.fold(&f6)
}

fn bjbt(csp: &CSP, t: Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)> {
    bj(&csp, bt(&csp, t))
}

fn bj_(_: &CSP, t: Node<(List<Assign>, ConflictSet)>) -> Node<(List<Assign>, ConflictSet)> {
    let f7 = |tp2, chs: List<Node<(List<Assign>, ConflictSet)>>| match tp2 {
        (a, ConflictSet::Known(cs)) => Node {
            lab: (a, ConflictSet::Known(cs)),
            children: chs.map(&Rc::new),
        },
        (a, ConflictSet::Unknown) => {
            let cs_ = ConflictSet::Known(combine(chs.clone().map(&|n| n.lab.clone()), List::Nil));
            if known_conflict(&cs_) {
                Node {
                    lab: (a, cs_),
                    children: List::Nil,
                }
            } else {
                Node {
                    lab: (a, cs_),
                    children: chs.map(&Rc::new),
                }
            }
        }
    };
    t.fold(&f7)
}

fn bjbt_(csp: &CSP, t: Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)> {
    bj_(csp, bt(csp, t))
}

fn fc(csp: &CSP, t: Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)> {
    domain_wipeout(
        csp,
        lookup_cache(csp, cache_checks(csp, empty_table(csp), t)),
    )
}

fn try_(
    n: i64,
    algorithm: fn(&CSP, Node<List<Assign>>) -> Node<(List<Assign>, ConflictSet)>,
) -> i64 {
    search(algorithm, &queens(n)).len() as i64
}

fn test_constraints_nofib(n: i64) -> List<i64> {
    [bt, bm, bjbt, bjbt_, fc]
        .map(&|x| try_(n, x))
        .iter()
        .map(|x| *x)
        .into()
}

fn main_loop(iters: u64, n: i64) {
    let res = test_constraints_nofib(n);
    if iters == 1 {
        println!("{}", res.head());
    } else {
        main_loop(iters - 1, n)
    }
}

fn main() {
    let mut args = std::env::args();
    args.next();
    let iters = args
        .next()
        .expect("Missing Argument iterations")
        .parse::<u64>()
        .expect("Iterations must be a number");
    let n = args
        .next()
        .expect("Missing Argument n")
        .parse::<i64>()
        .expect("m must be a number");
    main_loop(iters, n)
}
